### PR TITLE
Make <template> highlighting stable

### PR DESCRIPTION
**Pull request by NullVoxPopuli**
Using https://v5.chriskrycho.com/journal/ember-template-imports/part-2/ 
as some additional test cases, this PR aims to stabilize how `<template>` is interpreted in various scenarios. (thanks @chriskrycho!)

What needs to happen?
- [x] the js syntax for highlight.js needs to explicitly know about `<template>` and then defer to the glimmer grammar
  - [x] most of the work here (in hljs-glimmer) is tests, assuring that things are labeled correctly
  - [x] will need coordination with @joshgoebel of highlight.js to figure out what the best path forward is
    - precedence: https://github.com/tree-sitter/tree-sitter-javascript/pull/208 - this told tree-sitter to parse `<template>` specially, and the [injections here](https://github.com/tree-sitter/tree-sitter-javascript/blob/master/queries/injections.scm#L24-L32) declare how to defer highlighting to another grammar

### CODE REVIEW

**Comment by joshgoebel** [comment]
> will need coordination with @joshgoebel of highlight.js to figure out what the best path forward is

This is messy, but I'm open to suggestions.  Core grammars are not really intended to be extended from the outside like this.  But you could start by looking at Typescript...  How it starts with JS and then injects changes.  We could possibly add a few more named hooks if it's clear they were generally useful for these types of extensions.

You can do this at runtime as well by getting an existing language and accessing it's `rawDefinition` property - which will give you the original "uncompiled" data returned by the grammar function - that you can then inspect/walk to change programmatically to add your own tweaks - before injecting it back into HLJS as a new (or replacement language).

IE:

```
let grammar = HLJS.getLanguage("javascript").rawDefinition
// change grammar, add <template>, etc...
HLJS.registerLanguage("javascript", grammar)
```

For templates themselves you likely want `subLanguage`, see `xml` for examples.

**Comment by NullVoxPopuli** [comment]
Would you be open to API (and I don't know if this is possible, it's been a while), but to have something like:

```js
HLJS.registerInjection('javascript', () => {
  return {
    /* matcher thing */
    subLanguage: 'my sub language'
  }
}
```
and that would override anything the javascript language grammer would do?

**Comment by chriskrycho** [comment]
One note here—@NullVoxPopuli I would strongly suggest that we do *not* try to make it parse in cases where someone marks code as `js` or `javascript`, but instead support something like `gjs`, `gts`, and maybe `glimmer-js` and `glimmer-ts` respectively.

@joshgoebel this one is definitely quirky, but/and we may have further such embeddings. We’re thinking about whether and how we can design a more general "content tag" embedding system for JS—roughly an inversion of how HTML can have CSS and JS embedded via appropriate tags. Roughly something like this:

```
import executeQuery from 'some-graphql-engine';
import CoolStuff from './cool-stuff.gjs';

const Hero = <query lang='graphql'>
  {
    hero {
      name
      # Queries can have comments!
      friends {
        name
      }
    }
  }
</query>;

const Style = <style>
  .header {
    font-family: var(--sans);
  }

  .body {
    font-family: var(--serif);
  }
</style>;

function greeting(name?: string): string {
  return `Hello, ${name ?? 'world'}!`
}

<template lang='sveltelike'>
  <h1 class={Style.header}>{$args.title}</h1>
  <p class={Style.body}>{greeting($args.name)}</p>

  <CoolStuff @hero={executeQuery(Hero, { name: "Luke Skywalker" })} />
</template>
```

You can see how this is *very* similar to JSX, template tag-based embeddings, etc., but with some interesting differences in the trade-off space.

Net, we’d love to find a design that works well with how highlight.js thinks about embedded languages and which can generalize to support this.

**Comment by NullVoxPopuli** [comment]
> I would strongly suggest that we do not try to make it parse in cases where someone marks code as js or javascript, but instead support something like gjs, gts, and maybe glimmer-js and glimmer-ts respectively.

works for me :tada: the `setup` method provided by this library would only need to _not_ override the existing javascript language, and instead register glimmer-javascript, etc

File: src/index.js [hunk 1]
```diff
@@ -34,18 +34,109 @@ function registerJavaScriptInjections(hljs) {
...
+          response.ignoreMatch();
+
+          // eslint-disable-next-line no-useless-return
+          return;
+        }
+      }
+    },
+  };
+
+  js.contains.unshift({
```
Comment (NullVoxPopuli) [inline src/index.js:122]: @joshgoebel this is how I'm setting up the new injection.

I guess, what _I really want_ is some type-safe/jsdoc/intellisense-aware way or utility to get intellisense for what APIs I have access to, and what the shape of the grammar is that I'm modifying. even if it's a no-op wrapper function that adds jsdoc / type-info, that'd be a big help.

Comment (joshgoebel) [reply src/index.js:122]: The HLJS API itself has defined TypeScript type info, but there is no real "shape" to grammars (beyond the basic keys which are both typed and documented), and there are many different ways to write them... any type of extensions is going to have to hook into *very* specific points - and even then would possibly still be subject to the grammar changing out from under it.

Again, see typescript:

```
  // TS gets a simpler shebang rule than JS
  swapMode(tsLanguage, "shebang", hljs.SHEBANG());
  // JS use strict rule purposely excludes `asm` which makes no sense
  swapMode(tsLanguage, "use_strict", USE_STRICT);
```

Even to do those two simple changes the JS grammar first had to "tag" those two modes with "shebang" and "use_strict" - making it easy for the TS grammar to "find" them in the haystack that is the whole JS grammar... coming entirely from the outside you'd have to do your best to do this every release - or perhaps use some sort of heuristics to find the areas you wanted to patch in, etc.

**Comment by joshgoebel** [comment]
> and that would override anything the javascript language grammer would do?

It's not clear to me at all what `registerInjection` would do in your example - again you can replace entire grammars with `registerLanguage` already...

**Comment by joshgoebel** [comment]
` js.contains.unshift({`

You might actually get a lot of value out of that approach in many cases... if the grammar is largely top-level (which JS is).  Is there a reason that isn't working well enough for you?  Just inject the template matchers at the top (or bottom) of contains... and reregister the new language with a new key when you're done, as already mentioned earlier.

**Comment by NullVoxPopuli** [comment]
>  js.contains.unshift({
> Is there a reason that isn't working well enough for you? 

My main problem with this approach is that it feels like I'm reaching deep in to private API. 
I don't own the JS language grammar, and it's subject to change at any time, breaking anything and everything mucking about with its private API.

And since grammars don't include version information, I can't support a range of versions because implement if-branches in code -- but even doing would feels dirty because I'm messing with private API anyway.

**Comment by joshgoebel** [comment]
> My main problem with this approach is that it feels like I'm reaching deep in to private API.

Top-level `contains` is about as public as you can get.  Adding top-most rules is the safest thing you could do for most grammars.  If that works for your current use cases I'm not sure anything else is necessary.

If you want the most rock-solid safety possible, vendor the JS grammar or extend it directly vs trying to reuse the one we ship.

Doesn't your template stuff conflict with the JSX support though?

**Pull request merged**